src/abi_arm.cpp

```diff
@@ -0,0 +1,315 @@
+// Returns the corresponding LLVM type.
+static Type *isLegalHAType(jl_datatype_t *dt)
+{
+    // single- or double-precision floating-point type
+    Type* fp = get_llvm_fptype(dt);
+    if (fp)
+        return fp;
+
+    // 64- or 128-bit containerized vectors
+    if (isContainerizedVector(dt)) {
+        // TODO
+    }
+
+    // recursive application (composite types can contain other composites)
+    Type *base = NULL;
+    if (isLegalHA(dt, base))
+        return base;
+
+    return NULL;
+}
+
+// Check whether a type is a legal homogeneous aggregate. Returns the number of
+// members.
+static size_t isLegalHA(jl_datatype_t *dt, Type *&base) {
+    // Homogeneous aggregates are only used for VFP registers,
+    // so use that definition of legality (section 6.1.2.1)
+
+    if (jl_is_structtype(dt)) {
+        // ... with one to four Elements.
+        size_t members = jl_datatype_nfields(dt);
+        if (members < 1 || members > 4)
+            return 0;
+
+        base = NULL;
+        for (size_t i = 0; i < members; ++i) {
+            Type *T = isLegalHAType((jl_datatype_t*)jl_field_type(dt,i));
+            if (!T)
+                return 0;
+
+            if (!base)
+                base = T;
+            else if (base != T)
+                return 0;
+        }
+        return members;
+    }
+
+    return 0;
+}
+
+// Determine if an argument can be passed through a coprocessor register.
+//
+// All the out parameters should be default to `false`.
+static void classify_cprc(jl_datatype_t *dt, bool *vfp)
+{
+    // Based on section 6.1 of the Procedure Call Standard
+
+    // VFP: 6.1.2.1
+    // - A half-precision floating-point type.
+    // - A single-precision floating-point type.
+    // - A double-precision floating-point type.
+    if (get_llvm_fptype(dt)) {
+        *vfp = true;
+        return;
+    }
+
+    // - A 64-bit or 128-bit containerized vector type.
+    size_t vec_size = isContainerizedVector(dt);
+    if (vec_size == 8 || vec_size == 16) {
+        *vfp = true;
+        return;
+    }
+
+    // - A Homogeneous Aggregate
+    Type *base = NULL;
+    if (isLegalHA(dt, base)) {
+        *vfp = true;
+        return;
+    }
+}
+
+static void classify_return_arg(jl_value_t *ty, bool *reg, bool *vfpreg,
+                                bool *onstack, bool *need_rewrite)
+{
+    // Assume jl_is_datatype(ty) && !jl_is_abstracttype(ty)
+    jl_datatype_t *dt = (jl_datatype_t*)ty;
+
+    // Based on section 5.4 of the Procedure Call Standard
+
+    // VFP standard variant: see 6.1.2.2
+    //   Any result whose type would satisfy the conditions for a VFP CPRC is
+    //   returned in the appropriate number of consecutive VFP registers
+    //   starting with the lowest numbered register (s0, d0, q0).
+    classify_cprc(dt, vfpreg);
+    if (*vfpreg)
+        return;
+
+    // - A Half-precision Floating Point Type is returned in the least
+    //   significant 16 bits of r0.
+    if (dt == jl_float16_type) {
```

Comment:
I was trying to figure out how clang handles this but it (and also gcc) always complains that `__fp16` is not allowed as parameter or return type. Doesn't really hurt to keep this here though.